hip-0000-token-update-remove-keys.md

```diff
@@ -0,0 +1,127 @@
+---
+hip: 0000
+title: Remove existing keys from a Token
+author: Justyn Spooner (@justynjj)
+working-group: Justyn Spooner
+type: Standards Track
+category: Service
+needs-council-approval: <Yes | No>
+status: Draft
+created: 2022-08-05
+discussions-to: <a URL pointing to the official discussion thread>
+updated: <comma separated list of dates>
+requires: <HIP number(s)>
+replaces: <HIP number(s)>
+superseded-by: <HIP number(s)>
+---
+
+## Abstract
+
+This proposal will result in a modification to the `TokenUpdateTransaction` feature to allow the admin key to sign an update to remove itself and/or any other key (Wipe, KYC, Freeze, Pause, Supply, Fee Schedule) from a Token.
```

Comment:
>allow the admin key to sign an update to remove itself and/or any other key from a Token.

To me, this sounds like only the admin key can do this. It would be good if any key that is supplied may remove itself as well.

Reply:
That's a good point. I'll get that updated. Thanks for the feedback. 👍

Reply:
Adding another perspective: it would be preferable that only the admin can change or unset a sub key. Otherwise the admin key is not truly an admin but rather a manager-of-miscellaneous-fields-on-the-token-key. The admin should not have to play tug of war; it was granted the role of admin.

There's perhaps a corner case which could be of interest to discus: if the admin key is null and the kyc key is set, the kyc key can remove itself but not change itself to a new value.

Both of these support the least-privilege model.

Reply:
> Hello! I wanted to check and see what the status of this HIP currently is.
> 
> This is much needed functionality and would greatly empower creators.

still in community discussion. Seems there's still an outstanding issue around:

1. Should existing keys be required to sign when being rotated? 
2. Should a key be able to rotate itself without an ADMIN key?

Reply:
1. Yes - keys being rotated out should sign, which is consistent with the current implementation of the existing Hedera Network.
2. No - rotating keys is mutating the object, to be consistent with the current implementation of the existing Hedera Network, an admin key signals mutability, and therefore needs to participate in mutating an object configuration.

Reply:
Aside from "That's how Hedera Network does it", for what other possible reason(s) are the subkeys being required to sign? Is there some technical or non-technical reason for this design?

From a security perspective, that is a flawed approach; so would like to understand the above.

Reply:
Hedera got this right in the original design, except the overlooked the need to turn existing mutable objects into immutable ones, hence the need for this HIP.

Reply:
From what you're describing, it sounds like you're treating the supply key as the "supply key admin" and the trust of from the user is targeted to solely the supply admin and not the super root admin.

What you're proposing is uncommon. When compared to either the real world or to the tech domain. Other blockchains treat the admin as you say the "super root admin"; an admin doesn't need the permission from a subkey. In a real world comparison, a manager does not need an employees consent to lay them off (whereas in your assertion, the employee must consent to the manager's actions).

---

Giving a counter example to your example: if a supply key can hold the admin hostage for any reason at any point in time; then that should give the holders of that token concern. That the true power is controlled by the KYC key, the freeze key, the supply key, etc, since those keys can at any time disrupt the the token and the admin doesn't have unilateral control over them.

Reply:
I agree with @nkavian that the admin key is designed to be the super root user that has absolute control over the project. 

Example: You assign the supply key of your NFT project to one of your employees. This employee turns out to be malicious and leaves the project. When you don't have this admin super power, the project is stuck because the admin key can't rotate the supply key anymore. You need this type of super power initially to guarantee the safety over your project. Once you are confident that everything is running smoothly, you want to have the ability to remove the admin key. 

If the admin key can't change the supply key without its consent, we don't need this HIP I believe? @bugbytesinc

Reply:
The key that's being rotated doesn't need to sign. Here's a code example I have where I create an NFT with an admin key and supply key. Later, I change the supply key to a new random key. **Only the admin key had to sign the transaction,** the old and new supply key don't need to sign. This behavior is in line with the expected "super user" admin behavior. Let me know if this is correct? @bugbytesinc 

Here's the TX ID on testnet where you can see a different key for the supply key: https://hashscan.io/testnet/token/0.0.3934667 

Full code snippet (JS) is here: https://gist.github.com/michielmulders/7448e536734b80edff86de64d3e11e43

Reply:
> Nor the old, nor the new supply key need to sign for this transaction to be valid. So the above is in line with the existing behavior of the admin key having "super power" controls.

I stand corrected regarding admin key being a super root (there are other parts of the hedera system this is not the case, this is not one of them, so changing behavior would break backwards compatibility)  However the original hip (or this one) is still necessary if you want to make an component immutable after a period of bootstrapping.  Say, like freezing a supply of an NFT collection forever.

Reply:
just noting a few of the take aways here so we can reach consensus. 

1. ADMIN key should be able to remove/rotate *any* key without requiring that key's signature. 
Example: ADMIN key can remove SUPPLY key without requiring the SUPPLY key to sign the action. 

2. Any key can remove itself and (obviously) needs to sign the action. 

3.  When rotating the ADMIN key, both the old and new keys must sign the action (this is how Hedera current works)

....are we able to reach consensus on these 3 points?

Reply:
#2 is not backwards compatible.  ADMIN still should approve key swaps.  This is the model for mutability for the whole of hedera services.

Reply:
the issue I have here is that if no ADMIN key is set, then a key is immutable. There's plenty of reasons that a SUPPLY key can be set, but no ADMIN key set (minimal permissions). Should the SUPPLY key not then be able to remove itself to continue to reduce counterparty risk?

Reply:
I agree with Ashe's comment. If we want to follow the minimal permissions route, keys should be able to rotate itself. Adding to that, by being able to rotate themselves, we are not increasing the risk profile for the token.

Reply:
@bugbytesinc Let me know if you can agree with this? Happy to hear your further input into this matter. We are open to set up a call to discuss things and reach consensus 👐

Reply:
Perhaps @lbaird could provide insight into the reasoning for the original decision keep objects immutable that don't have admin keys.  Should a token that has no ADMIN key be able to have the SUPPLY key remove itself so there is no longer a SUPPLY key at all?